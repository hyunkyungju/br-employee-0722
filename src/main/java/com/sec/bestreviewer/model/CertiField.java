package com.sec.bestreviewer.model;

import lombok.Getter;

@Getter
public enum CertiField {
    ADV("adv"),
    PRO("pro"),
    EX("ex");

    private final String fieldName;

    CertiField(String fieldName) {
        this.fieldName = fieldName;
    }

    public int getIndex() {
        return ordinal();
    }
}
