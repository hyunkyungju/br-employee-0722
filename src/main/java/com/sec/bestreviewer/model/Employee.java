package com.sec.bestreviewer.model;


import com.sec.bestreviewer.util.FullNameValidator;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NonNull;

import java.util.Objects;

@Getter
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class Employee {
    @NonNull private final String employeeNumber;
    @NonNull private String name;
    @NonNull private String careerLevel;
    @NonNull private String phoneNumber;
    @NonNull private String birthday;
    @NonNull private String certi;

    // 복사 생성자 추가
    public Employee(Employee other) {
        this.employeeNumber = other.employeeNumber;
        this.name = other.name;
        this.careerLevel = other.careerLevel;
        this.phoneNumber = other.phoneNumber;
        this.birthday = other.birthday;
        this.certi = other.certi;
    }

    public static Employee of(String employeeNumber, String fullName, String careerLevel, String phoneNumber, String birthday, String certi) {
        if (!FullNameValidator.isValidFullName(fullName)) {
            throw new IllegalArgumentException("fullName 형식이 올바르지 않습니다.");
        }
        return new Employee(employeeNumber, fullName, careerLevel, phoneNumber, birthday, certi);
    }

    public String getFirstName() {
        return name.split(" ")[0];
    }

    public String getLastName() {
        return name.split(" ")[1];
    }

    @Override
    public boolean equals(Object object) {
        if (this == object) {
            return true;
        }
        if (object == null || getClass() != object.getClass()) return false;
        final Employee employee = (Employee) object;
        return Objects.equals(employeeNumber, employee.employeeNumber);
    }

    @Override
    public int hashCode() {
        return Objects.hash(employeeNumber);
    }

    public String getFieldValue(EmployeeField field) {
        switch (field) {
            case EMPLOYEE_NUMBER:
                return employeeNumber;
            case NAME:
                return name;
            case CAREER_LEVEL:
                return careerLevel;
            case PHONE_NUMBER:
                return phoneNumber;
            case BIRTHDAY:
                return birthday;
            case CERTI:
                return certi;
            default:
                return null;
        }
    }

    public void setFieldValue(EmployeeField employeeField, String value) {
        switch (employeeField) {
            case EMPLOYEE_NUMBER:
                // employeeNumber는 final이므로 set 불가
                break;
            case NAME:
                this.name = value;
                break;
            case CAREER_LEVEL:
                this.careerLevel = value;
                break;
            case PHONE_NUMBER:
                this.phoneNumber = value;
                break;
            case BIRTHDAY:
                this.birthday = value;
                break;
            case CERTI:
                this.certi = value;
                break;
            default:
                break;
        }
    }
}
