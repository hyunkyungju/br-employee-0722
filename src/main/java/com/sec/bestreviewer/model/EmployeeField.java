package com.sec.bestreviewer.model;

import lombok.Getter;

import java.util.Arrays;

@Getter
public enum EmployeeField {
    EMPLOYEE_NUMBER("employeeNum"),
    NAME("name"),
    CAREER_LEVEL("cl"),
    PHONE_NUMBER("phoneNum"),
    BIRTHDAY("birthday"),
    CERTI("certi");

    private final String fieldName;

    EmployeeField(String fieldName) {
        this.fieldName = fieldName;
    }

    public static EmployeeField parseField(String input) {
        return Arrays.stream(EmployeeField.values())
                .filter(field -> field.getFieldName().equalsIgnoreCase(input))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("Unknown field: " + input));
    }
}
