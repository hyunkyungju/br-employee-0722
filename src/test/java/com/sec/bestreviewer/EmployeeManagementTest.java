package com.sec.bestreviewer;

import com.sec.bestreviewer.model.CertiField;
import com.sec.bestreviewer.model.Employee;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;


import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.text.Format;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class EmployeeManagementTest {
    @Test
    public void testArgumentsEmptyInputFile() throws Exception {
        String[] args = {"./src/test/java/com/sec/bestreviewer/empty.txt", "./src/test/java/com/sec/bestreviewer/output.txt"};

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        File file = new File("./src/test/java/com/sec/bestreviewer/output.txt");

        assertTrue(file.exists());
    }

    @Test
    public void testArgumentsWrongArgsCount() throws Exception {
        String[] args = {"input.txt"};

        EmployeeManagement employeeManagement = new EmployeeManagement();

        Assertions.assertThrows(IllegalArgumentException.class, () -> {
            employeeManagement.run(args);
        });
    }

    @Test
    public void testArgumentsNotExistInputFile() throws Exception {
        String[] args = {"notexist.txt", "output.txt"};

        EmployeeManagement employeeManagement = new EmployeeManagement();

        Assertions.assertThrows(IllegalArgumentException.class, () -> {
            employeeManagement.run(args);
        });
    }

    @Test
    @Disabled
    public void integrationTest() throws Exception {
        final String outputFileName = "./src/test/java/com/sec/bestreviewer/integration_test_output.txt";
        final String inputFileName = "./src/test/java/com/sec/bestreviewer/integration_test_input.txt";

        String[] args = {inputFileName, outputFileName};
        File outputFile = new File(outputFileName);

        EmployeeManagement employeeManagement = new EmployeeManagement();

        employeeManagement.run(args);

        Approvals.verify(outputFile);
    }

    @Test
    public void testEmployeeNullNameCreation() throws Exception {
        Assertions.assertThrows(NullPointerException.class, () -> {
            Employee employee =
                    Employee.of("15087654",
                            null,
                            "CL1",
                            "010-1234-5678",
                            "20190101",
                            CertiField.ADV.getFieldName());
        });
    }

    @Test
    public void testEmployeeNullCareerLevelCreation() throws Exception {
        Assertions.assertThrows(NullPointerException.class, () -> {
            Employee employee =
                    Employee.of("15087654",
                            "SEO KFI",
                            null,
                            "010-1234-5678",
                            "20190101",
                            CertiField.ADV.getFieldName());
        });
    }

}

