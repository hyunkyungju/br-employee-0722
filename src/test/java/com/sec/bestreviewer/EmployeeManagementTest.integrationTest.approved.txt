SCH,9
SCH,NONE
SCH,02712462,UAFBOWU HONG,CL3,010-9054-6560,19780825
SCH,09752663,VWCGM OH,CL3,010-6288-1811,19860720
SCH,94526445,VLPDHI LEE,CL4,010-7975-9766,19721111
SCH,NONE
SCH,NONE
SCH,NONE
SCH,99398134,JYFMBMT OH,CL4,010-5269-9657,19730321
SCH,NONE
SCH,02712462,UAFBOWU HONG,CL3,010-9054-6560,19780825
SCH,NONE
SCH,NONE
SCH,NONE
SCH,02712462,UAFBOWU HONG,CL3,010-9054-6560,19780825
SCH,03717863,NSESFBLRK KIM,CL3,010-4108-4007,19770502
SCH,7
SCH,NONE
SCH,93728322,QSJYRB HONG,CL4,010-3602-4064,19690603
SCH,03249872,TYLNXWQN OH,CL3,010-7910-1669,19820828
